core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -0,0 +1,384 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Planner rule that matches a
+ * {@link org.apache.calcite.rel.core.Join}
+ * and expands OR clauses in join conditions.
+ *
+ * <p>this rule would expand the OR condition into
+ * two separate join conditions, allowing the optimizer
+ * to handle these conditions more effectively.
+ */
+@Value.Enclosing
+public class JoinConditionOrExpansionRule
+    extends RelRule<JoinConditionOrExpansionRule.Config>
+    implements TransformationRule {
+
+  /** Creates an JoinConditionExpansionOrRule. */
+  protected JoinConditionOrExpansionRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Join join = call.rel(0);
+    RelBuilder relBuilder = call.builder();
+
+    RelNode converted;
+    switch (join.getJoinType()) {
+    case INNER:
+      converted = expandInnerJoin(join, relBuilder);
+      break;
+    case ANTI:
+      converted = expandAntiJoin(join, relBuilder);
+      break;
+    case LEFT:
+      converted = expandLeftOrRightJoin(join, true, relBuilder);
+      break;
+    case RIGHT:
+      converted = expandLeftOrRightJoin(join, false, relBuilder);
+      break;
+    case FULL:
+      converted = expandFullJoin(join, relBuilder);
+      break;
+    default:
+      return;
+    }
+
+    if (converted == null) {
+      return;
+    }
+    call.transformTo(converted);
+  }
+
+  private boolean skip(Join join) {
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+    return orConds.size() <= 1;
+  }
+
+  private List<RexNode> splitCond(Join join) {
+    final RexBuilder builder = join.getCluster().getRexBuilder();
+    final List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+    final int leftFieldCount = join.getLeft().getRowType().getFieldCount();
+
+    final List<RexNode> result = new ArrayList<>();
+    final List<RexNode> otherBuffer = new ArrayList<>();
+
+    for (RexNode cond : orConds) {
+      if (isUsefulCond(cond, leftFieldCount)) {
+        if (!otherBuffer.isEmpty()) {
```

Comment:
This small fragment could be a utility function "creatDisjunction".
I wonder whether this exists already somewhere.

Reply:
Done!